src/main/java/org/broadinstitute/hellbender/tools/spark/sv/RunSGANaivelySpark.java

```diff
@@ -0,0 +1,485 @@
+import org.broadinstitute.hellbender.engine.spark.GATKSparkTool;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.io.IOUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+import scala.Tuple2;
+import scala.Tuple3;
+
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+// TODO: choose which parameters allowed to be tunable
+// TODO: if throws, would temp files be cleaned up automatically?
+// TODO: most robust way to check OS in method SGASerialRunner (Mac doesn't support multi-thread mechanism in SGA)?
+@CommandLineProgramProperties(
+        summary        = "Program to call SGA for performing local assembly and outputs assembled contigs.",
+        oneLineSummary = "Perform SGA-based local assembly on fasta files on Spark",
+        programGroup   = StructuralVariationSparkProgramGroup.class)
+public final class RunSGANaivelySpark extends GATKSparkTool {
+    private static final long serialVersionUID = 1L;
+
+    @Argument(doc       = "Absolute path to SGA installation.",
+              shortName = "sgaPath",
+              fullName  = "fullPathToSGA",
+              optional  = false)
+    public String pathToSGA = null;
+
+    @Argument(doc       = "A path to a directory to write results to.",
+              shortName = "outDir",
+              fullName  = "outputDirectory",
+              optional  = false)
+    public String outputDir = null;
+
+    @Argument(doc       = "Number of threads to use when running sga.",
+              shortName = "t",
+              fullName  = "threads",
+              optional  = true)
+    public int threads = 1;
+
+    @Argument(doc       = "To run k-mer based read correction, filter and duplication removal in SGA or not, with default parameters.",
+              shortName = "correct",
+              fullName  = "correctNFilter",
+              optional  = true)
+    public boolean runCorrectionSteps = false;
+
+    @Override
+    public boolean requiresReads(){
+        return true;
+    }
+
+    @Override
+    public void runTool(final JavaSparkContext ctx){
+
+        final JavaRDD<GATKRead> reads = getReads();
+        // TODO: experimental interface to caller, should be getting this from JavaSparkContext
+        final JavaPairRDD<Long, Iterable<GATKRead>> readsAroundBreakpoints = reads.groupBy( read -> 1L);
+
+        final Path sgaPath = Paths.get(pathToSGA);
+
+        // map paired reads to FASTA file (temp files live in temp dir that cleans self up automatically) then feed to SGA for assembly
+        final JavaPairRDD<Long, SGAAssemblyResultType> assembledContigs = readsAroundBreakpoints.mapToPair(RunSGANaivelySpark::convertReadsToFASTQ)
```

Comment:
Have you tested this on the cluster? I am not sure if it will work the way it is currently structured. (If you have tested it, apologies, I must have a misunderstanding about how things work here).

Here's the issue:

The first `mapToPair` will create temporary directories with FASTQ files in them. Won't it create these on the local filesystems? The fact that the lists of these directories is then kept in an RDD means that they should be viewed in the distributed context, so if the RDD is shuffled or whatnot you might end up with RDD entries that have paths to files that are not on the nodes that are processing that RDD element. It could be that since you're never shuffling or re-partitioning the RDD this won't be a problem, but my feeling is that it's best not to have links to node-local resources (that aren't accessible to all nodes) in an RDD.

I'd suggest combining the writing of FASTQ files and the running of SGA into a single step.

Happy to discuss in person.